### PR TITLE
[browser][websocket] Throw OperationCanceledException on connect

### LINKED ISSUE #44720: [browser][websocket] Canceling token source before connect should throw `OperationCanceledException`
The WebSocketClient should throw an `OperationCanceledException` instead of trying to connect and then cancelling.

```
            using (var clientSocket = new ClientWebSocket())
            {
                var cts = new CancellationTokenSource();
                cts.Cancel();
                Task t = clientSocket.ConnectAsync(new Uri("ws://" + Guid.NewGuid().ToString("N")), cts.Token);
            }
```

### PR DESCRIPTION
**Pull request by kjpou1**
if cancel was requested before.

This fixes the test

```
        public async Task ConnectAsync_CancellationRequestedBeforeConnect_ThrowsOperationCanceledException()
        {
            using (var clientSocket = new ClientWebSocket())
            {
                var cts = new CancellationTokenSource();
                cts.Cancel();
                Task t = clientSocket.ConnectAsync(new Uri("ws://" + Guid.NewGuid().ToString("N")), cts.Token);
                await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
            }
        }
```

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.cs#L249

Fix https://github.com/dotnet/runtime/issues/44720

Part of https://github.com/dotnet/runtime/issues/44666

Test was failing with the following error:

```
[FAIL] System.Net.WebSockets.Client.Tests.ConnectTest.ConnectAsync_CancellationRequestedBeforeConnect_ThrowsOperationCanceledException
  info: Assert.Throws() Failure
  info: Expected: typeof(System.OperationCanceledException)
  info: Actual:   typeof(System.Net.WebSockets.WebSocketException): Unable to connect to the remote server
  info: ---- System.Net.WebSockets.WebSocketException : Unable to connect to the remote server
  info: -------- System.Net.WebSockets.WebSocketException : An internal WebSocket error occurred. Please see the innerException, if present, for more details.
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)
  info:    at System.Net.WebSockets.WebSocketHandle.ConnectAsync(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
  info: ----- Inner Stack Trace -----
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)
  info:    at System.Net.WebSockets.WebSocketHandle.ConnectAsync(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
  info: ----- Inner Stack Trace -----
  info:    at System.Net.WebSockets.BrowserWebSocket.ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List`1 requestedSubProtocols)

```

See https://github.com/dotnet/runtime/pull/45470 for more tests

### CODE REVIEW

**Comment by stephentoub** [comment]
> This fixes the test

I understand that this change avoids doing extra work in case cancellation was already requested, but can you elaborate on how it fixes the test?  What was happening before this change?

**Comment by kjpou1** [comment]
Updated the PR comments to include the error message.

**Comment by stephentoub** [comment]
> Test was failing with the following error:

But doesn't that suggest there's still a problem here?  What happens if cancellation is requested while the connect is in flight?  That should still produce an OperationCanceledException, but the error you pasted suggests it isn't.

**Comment by lewing** [comment]
The test change is a race on anything with threads so it should be changed back and and a threadless platform test added.

**Comment by kjpou1** [comment]
I am not finding it racey.  I will take a closer look but it seems to be working fine on my local machine.

**Comment by lewing** [comment]
I meant in theory it might be able to fail on the host lookup before the cancellation, upon reflection that is unlikely, I'm fine either way.

**Comment by kjpou1** [comment]
I have made multiple tests with these modifications against a local server as well as against a public websocket test echo server with no problems.

The new test also works as expected.  The original tests tests against a server that was not found but testing against a live server where the connection actually exists and succeeds also works.

```
            using (var clientSocket = new ClientWebSocket())
            {
                var cts = new CancellationTokenSource();
                //Task t = clientSocket.ConnectAsync(new Uri("ws://" + Guid.NewGuid().ToString("N")), cts.Token);
                Task t = clientSocket.ConnectAsync(new Uri("wss://echo.websocket.org"), cts.Token);
                cts.Cancel();
                await Assert.ThrowsAnyAsync<OperationCanceledException>(() => t);
            }
```

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 1]
```diff
@@ -13,6 +13,10 @@ public static class WebSocketHelper
     {
         private static readonly Lazy<bool> s_WebSocketSupported = new Lazy<bool>(InitWebSocketSupported);
         public static bool WebSocketsSupported { get { return s_WebSocketSupported.Value; } }
+        // On browser the JavaScript host object 'WebSocket' is only available when DOM is supported
+        // If DOM is not available then a JavaScript error is thrown
+        // --- System.Runtime.InteropServices.JavaScript.JSException : JavaScript host object 'WebSocket' not found.
+        public static bool WebSocketsSupportedInBrowser { get { return !PlatformDetection.IsBrowser || (PlatformDetection.IsBrowser && PlatformDetection.IsBrowserDomSupported); } }
```
Comment (lewing) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:19]: I don't understand this new check.  WebSocketsSupported supported checks for PNSE which is what we should throw if the platform doesn't support it (and the other tests suggest that we do). If the goal is to only run this test on browser-wasm use PlatformSpecific

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs [hunk 2]
```diff
@@ -61,7 +62,8 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken, Cli
 
                 Abort();
 
-                if (exc is WebSocketException)
+                if (exc is WebSocketException ||
+                    (exc is OperationCanceledException && cancellationToken.IsCancellationRequested))
```
Comment (lewing) [inline src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:66]: I'm not sure what the optimal/expected behavior here is but this check should probably match the BrowserWebSocket case

**Review by lewing — COMMENTED** [review]
The fact checks need work.

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs (hunk 1 above)
Comment (lewing) [reply src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:19]: To expand on my previous comment, if we aren't properly throwing PNSE when the JS runtime doesn't support `WebSocket` we should fix that.  If you want to skip the test on platforms other than Browser use [PlatformSpecific(TestPlatforms.Browser)]. The combination of the two should be enough?

**Comment by lewing** [comment]
@kjpou1 I've pushed a rough pass at what I was suggesting because I realized we've discussed the complications before.

**Comment by lewing** [comment]
And now I'm realizing the platform check was probably written that way because of the old options throwing.  If browser is the only unsupported case we simplify it a lot more.

**Comment by kjpou1** [comment]
Agree and agree.  The reason we did not throw if WebSocket was not found was so that we could do rudimentary tests with the javascript engines.  You are right that now that we have the tests stood up via browser we can now do it this way.

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs (hunk 2 above)
Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs]: This should already be resolved via commits below by yourself

**Comment by kjpou1** [comment]
@lewing ping.  Can you review this again?

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs [hunk 3]
```diff
@@ -14,7 +14,11 @@ internal sealed class WebSocketHandle
         public WebSocket? WebSocket { get; private set; }
         public WebSocketState State => WebSocket?.State ?? _state;
 
-        public static ClientWebSocketOptions CreateDefaultOptions() => new ClientWebSocketOptions();
+        public static ClientWebSocketOptions CreateDefaultOptions()
+        {
+            BrowserWebSocket.ThrowIfPlatformNotSupported();
```
Comment (marek-safar) [inline src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: Is this the right place for the throw? Which calls are you trying to protect?

File: src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs [hunk 4]
```diff
@@ -13,7 +13,6 @@ public static class WebSocketHelper
     {
         private static readonly Lazy<bool> s_WebSocketSupported = new Lazy<bool>(InitWebSocketSupported);
         public static bool WebSocketsSupported { get { return s_WebSocketSupported.Value; } }
-
```
Comment (marek-safar) [inline src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs:16]: Revert this change

Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs]: addressed

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs (hunk 3 above)
Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: I would say yes and not protecting any calls that I know of.  It allows for checking in the tests if we are able to run the tests for browser or not.

Comment (marek-safar) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: Then why not to do the checks in the tests and exclude the tests on configurations which don't support websockets?

Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: The tests check for PlatformNotSupported

Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: Also, the original code did the checks in the tests and it was suggested to do it this way.  Whichever way is ok with me just let me know.

Comment (lewing) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:19]: I don't think it makes any sense at this point to have WebSocketsSupported be true on v8 and then add another condition for when it is really true on browser.  If browser is the only platform that doesn't support WebSockets can't the condition can just become PlatformDetection.IsNotBrowser || PlatformDetection.IsBrowserDomSupported.  If there isn't a easy way to express that in attributes we can either stick it in InitWebSocketSupported or add another PlatformDetection value, something like IsClientNetworkingSupported which would also cover the outgoing http case.

Comment (lewing) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs]: WebSocketsSupported is used other places so I moved the check to the tests only, if we ever intend to officially support wasm-browser on !browser we'll have more places to fix than just here.

Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs]: Thanks Larry

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs [hunk 5]
```diff
@@ -61,11 +62,14 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken, Cli
 
                 Abort();
 
-                if (exc is WebSocketException)
-                {
-                    throw;
+                switch (exc) {
+                    case WebSocketException:
+                    case PlatformNotSupportedException:
```
Comment (marek-safar) [inline src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs:67]: Is this case still needed?

Comment (kjpou1) [reply src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Browser.cs]: Addressed in recent commit

**Comment by marek-safar** [comment]
Looks good to me but will wait on @lewing to approve

**Review by lewing — APPROVED** [review]
